### PR TITLE
Update multinomial.py

### PR DESCRIPTION
**Pull request by mingxuan-yi**
The previous calculation of log_prob is unstable if the probs of Multinomial are close to zeros, the math_ops.log will result in nan values.

### CODE REVIEW

File: tensorflow/python/ops/distributions/multinomial.py [hunk 1]
```diff
@@ -260,7 +261,7 @@ def _log_prob(self, counts):
 
   def _log_unnormalized_prob(self, counts):
     counts = self._maybe_assert_valid_sample(counts)
-    return math_ops.reduce_sum(counts * math_ops.log(self.probs), -1)
+    return -nn_ops.softmax_cross_entropy_with_logits(labels=counts, logits=self.logits)
```
Comment (ebrevdo) [inline tensorflow/python/ops/distributions/multinomial.py]: there is a known bug wherein softmax_cross_entropy_with_logits does not backprop into its labels argument; so you cannot backprop into counts using this function.

**Review by ebrevdo — COMMENTED** [review]
stability is nice; maybe there is a helper function in disribution_utils you can use?

**Comment by YiMX** [comment]
Hi, after some testing, I found that log_prob computed from math_ops.reduce_logsumexp with logits can be stable and counts is trainable in this case.

**Comment by drpngx** [comment]
@YiMX did you find a helper function in `distribution_utils`?

**Comment by sb2nov** [comment]
@ebrevdo ping

File: tensorflow/python/ops/distributions/multinomial.py [hunk 2]
```diff
@@ -260,7 +260,8 @@ def _log_prob(self, counts):
 
   def _log_unnormalized_prob(self, counts):
     counts = self._maybe_assert_valid_sample(counts)
-    return math_ops.reduce_sum(counts * math_ops.log(self.probs), -1)
+    logsumexp = math_ops.reduce_logsumexp(self.logits, -1, keep_dims=True)
```
Comment (ebrevdo) [inline tensorflow/python/ops/distributions/multinomial.py]: this doesn't _seem_ to be the same calculation.  sorry if i'm being thick, can you show how this calculation is equivalent mathematically to the old one?

**Review by ebrevdo — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/distributions/multinomial.py (hunk 2 above)
Comment (YiMX) [reply tensorflow/python/ops/distributions/multinomial.py]: default
```
from tensorflow.python.ops import math_ops
import tensorflow as tf

with tf.Session() as sess:
    counts = [1., 2., 3.]
    logits = [-1., 2., 3.]
    probs = tf.exp(logits) / tf.reduce_sum(tf.exp(logits))
    
    # The previous one
    print('The previous one', sess.run(math_ops.reduce_sum(counts * math_ops.log(probs), -1)))
    
    # Minus cross entropy
    print('Minus Cross entropy', sess.run(-tf.nn.softmax_cross_entropy_with_logits(labels=counts, logits=logits)))
    
    # The current one
    logsumexp = math_ops.reduce_logsumexp(logits, -1, keep_dims=True)
    print('The current one',sess.run(math_ops.reduce_sum(counts * (logits - logsumexp), -1)))

    
    # Let some of probs be close to zero
    counts = [1., 2., 3.]
    logits = [-1000., 2000., 2000.]
    probs = tf.exp(logits) / tf.reduce_sum(tf.exp(logits))
    
    # The previous one
    print('The previous one',sess.run(math_ops.reduce_sum(counts * math_ops.log(probs), -1)))
    
    # Minus cross entropy
    print('Minus Cross entropy', sess.run(-tf.nn.softmax_cross_entropy_with_logits(labels=counts, logits=logits)))
    
    # The current one
    logsumexp = math_ops.reduce_logsumexp(logits, -1, keep_dims=True)
    print('The current one',sess.run(math_ops.reduce_sum(counts * (logits - logsumexp), -1)))
```
**The above returns:**

('The previous one', -7.9593754)
('Minus Cross entropy', -7.9593763)
('The current one', -7.9593763)
('The previous one', nan)
('Minus Cross entropy', -3004.1587)
('The current one', -3004.1587)

**Review by ebrevdo — APPROVED** [review]

**Comment by jvdillon** [comment]
Hi folks. It looks like this issue was already spotted and we have a fix about to go in.  Our fix uses tf.nn.log_softmax which is slightly more efficient than this PR.  Again, my apologies for not intervening sooner (I was only just made aware of this PR).

**Comment by gunan** [comment]
@jvdillon is the other fix merged? Should we close this PR?

**Comment by YiMX** [comment]
@gunan  It was merged, I will close this PR.
